### DISCUSSION TITLE
[xpath engine] return the search url when no result is found

### DISCUSSION THREAD

**Discussion by qaristote** [searx/searx#3041] [Development]
Many websites skip the result page when the user searches for something that has an exact match. For example, when searching for Searx on [AlternativeTo](https://alternativeto.net/), the result page https://alternativeto.net/browse/search?q=searx directly redirects the user to https://alternativeto.net/software/searx/. 

This is a problem when trying to add the website as a search engine using xpath because the page to which the result page redirects to may not be parsed the same way as the result page itself, leading to a false-negative "no results found" error.

For example, when adding AlternativeTo to Searx using the configuration
```
- name : alternativeto
  engine : xpath
  paging : True
  search_url : https://alternativeto.net/browse/search?q={query}&p={pageno}
  results_xpath : //article[@class="row app-list-item"]/div[@class="col-xs-10 col-sm-10 col-md-11 col-lg-offset-1 col-lg-11"]
  url_xpath : ./h3/a/@href
  title_xpath : ./h3/a
  content_xpath : ./div[@class="itemDesc read-more-box"]/p[@class="text"]
```
the `alternativeto` engine will return results when searching for "sear" but will not return any result when searching for "searx".

My current solution is to add an invisible whitespace at the end of the queries, for example using 
```
search_url : https://alternativeto.net/browse/search?q={query}%E2%80%8E&p={pageno}
```
for AlternativeTo, but this is not ideal as it doesn't return the exact same results as without the invisible whitespace and requires that the `%E2%80%8E` string be copied in each xpath engine entry.

What I think would be a great solution would be that the xpath engine return the search url given in `search_url` whenever no result is found when parsing the page. The title and content of the result may then simply be the title and description of the webpage found in the HTTP header.
Of course this may only be added as an option that is disabled by default.

What do you think ?

**Comment by allendema** [comment]
Works well - you can just add ```/?ignoreExactMatch=true``` to search_url variable like below.
This way it will ignore popular matches.


```diff
  - name : alternativeto
    shortcut : ato
    engine : xpath
    timeout : 4.0
    paging : True
    search_url : https://alternativeto.net/browse/search/?ignoreExactMatch=true&q={query}&p={pageno}
    results_xpath : //article[@class="row app-list-item"]/div[@class="col-xs-10 col-sm-10 col-md-11 col-lg-offset-1 col-lg-11"]
    url_xpath : ./h3/a/@href
    title_xpath : ./h3/a
    content_xpath : ./div[@class="itemDesc read-more-box"]/p[@class="text"]
    disabled : True
```
Notice that ```shortcut```, ```timeout``` and ```disabled``` are addded but you can change them to your liking.

**Reply by qaristote** [reply]
That is indeed a great solution in the case of AlternativeTo, but it's not a standard solution, is it ? For example in the case of [Emojipedia](https://emojipedia.org/), it doesn't work : https://emojipedia.org/search/?ignoreExactMatch=true&q=🤔 immediately redirects to https://emojipedia.org/thinking-face/. 

So unless there is a way to know what keyword to use each time you want to add a new engine (?), I still think a more general solution should be available.

**Reply by allendema** [reply]
Correct, not a standard solution because diffrent engines use different keys (like ```/?ignoreExactMatch```).

>> What I think would be a great solution would be that the xpath engine return the search url given in search_url whenever no result is found when parsing the page.

This means that every result ill have the same result url? For some enignes it may be a good idea.

You can also look ```max_redirects``` and ```soft_max_redirects``` from the [Docs](http://searx.github.io/searx/dev/engine_overview.html?highlight=max_redirects#id12) and [how they are used in Searx.](https://github.com/searx/searx/search?q=max_redirects&type=commits)

**Reply by qaristote** [reply]
> This means that every result ill have the same result url? 

What I had in mind was that as long as some results are found, the engine should behave as it already does (in particular each result still has its own url).
But when no results are found, for example because an exact match has been found and the result page cannot be parsed using the given xpaths, the engine should return a single result whose url is the response's url (so possibly a 
 url resulting from a redirection).

Something like this :
```
def results(resp) : 
    results = parse_results(resp)
    if results == [] : 
        url = resp.url
        html = bs4.BeautifulSoup(html)
        title = html.head.title.text
        content = "" #TODO
        results = [ { 'url' : url, 'title' : title, 'content' : content } ]
     return results
```
But this is assuming the use of BeautifulSoup, and I don't know if there's a universal way to get the description of the content of the webpage in HTML. And of course I guess this also requires some error handling.
